File: src/sandbox/ai_prj1/Graph.ts
import PriorityQueue from 'ts-priority-queue'
import { Observable, voidFn } from './Observable'
import {delay, IEdge, IUIMessage, IVertex, TRACE_COLOR, VISITED_COLOR, CURRENT_COLOR } from './shared'

const COLS = 10
const ROWS = 10
const DELAY = 50

const getSquarePos = (v: IVertex) => {
    const x = Math.floor(v.squareID / ROWS)
    const y = v.squareID % ROWS

    return {x,y}
}


// Heuristic function used in A*
const heur = (a: IVertex, b: IVertex) => {
    const aPos = getSquarePos(a)
    const bPos = getSquarePos(b)

    const side1 = Math.abs(aPos.x - bPos.x)
    const side2 = Math.abs(aPos.y - bPos.y)

    return Math.sqrt(side1*side1 + side2*side2)
}

type Algorithm = 'dijkstra' | 'astar'

export class Graph {
    adjList: {[key:number]: IEdge[]} = {}

    vertexMap: {[key:number] : number} = {}

    constructor(private observer: Observable<IUIMessage>) {}

    subscribe(fn: voidFn<IUIMessage>) {
        this.observer.subscribe(fn)
    }

    notifyUI(vertexID: number, color: string) {
        this.observer.notify({
            squareID: this.vertexMap[vertexID],
            color
        })
    }

    addEdge(from: number, to: number, distance: number) {
        if(!this.adjList[from]) {
            this.adjList[from] = []
        }
        if(!this.adjList[to]) {
            this.adjList[to] = []
        }

        this.adjList[from].push({from,to,distance})
        this.adjList[to].push({from,to,distance})
    }

    addEdges(edges: IEdge[]) {
        for(let {from,to,distance} of edges) {
            this.addEdge(from,to,distance)
        }
    }

    addVertex(v: IVertex) {
        this.vertexMap[v.vertexID] = v.squareID
    }

    addVertices(vertices: IVertex[]) {
        for(let vertex of vertices) {
            this.addVertex(vertex)
        }
    }

    printPath(end: number, parent: number[]) {
        const stack = []
        while(end !== -1) {
            stack.push(end)
            end = parent[end]
        }
        const path:number[] = []
        for(let i = stack.length - 1; i >= 0; i--) {
            path.push(stack[i])
            this.notifyUI(stack[i],TRACE_COLOR)
        }
        return path
    }

    countVisited = (visited: boolean[]) => {
        let count = 0
        for(let v of visited) {
            count++
        }
        return count
    }


    shortestPath(start:number, goal: number, searchType: Algorithm) {
        switch(searchType) {
            case 'dijkstra':
                this.dijkstra(start,goal)
                break
            case 'astar':
                this.astar(start,goal)
                break
            default:
                console.log("Error, please specify a valid search type")
        }
    }

    async dijkstra(start:number, goal: number) {
        let count = 0
        const dist = new Array(Object.keys(this.adjList).length).fill(Infinity)
        const parent = new Array(Object.keys(this.adjList).length).fill(-1)
        dist[start] = 0

        const agenda = new PriorityQueue<IEdge>({comparator: (a,b) => a.distance - b.distance})
        const selfEdge = {from:start, to:start, distance: 0}
        agenda.queue(selfEdge)

        while(agenda.length > 0) {
            const currEdge = agenda.dequeue()

            // Update UI
            this.notifyUI(currEdge.from,CURRENT_COLOR)

            await delay(DELAY)

            const currNode = currEdge.to
            if(currEdge.distance > dist[currNode]) {
                this.notifyUI(currEdge.from,VISITED_COLOR)
                continue
            }

            for(let edge of this.adjList[currNode]) {
                count++
                const {from,to, distance} = edge
                if(dist[currNode] + distance < dist[to]) {
                    dist[to] = dist[currNode] + distance
                    agenda.queue({from,to,distance: dist[to]})
                    parent[to] = currNode
                }
            }

            this.notifyUI(currEdge.from,VISITED_COLOR)
        }

        if(isFinite(dist[goal])) {
            // console.log("DIST: ", dist)
            console.log("PATH FOUND WITH COST: ", dist[goal])
            console.log("PATH: " + this.printPath(goal,parent))
            console.log("Dijkstra Visited nodes count: ", count)
        } else {
            console.log("DIJKSTRA NO SOLUTION")
        }
    }

    async astar(start:number, goal: number) {
        const agenda = new PriorityQueue<IEdge>({comparator: (a,b) => a.distance - b.distance})
        let visited: boolean[] = []
        let parent: number[] = []
        let gScore: number[] = []
        let fScore: number[] = []
        for(let i = 0; i < Object.keys(this.adjList).length; i++) {
            visited[i] = false
            parent[i] = -1
            gScore[i] = Infinity
            fScore[i] = Infinity
        }

        const selfEdge = {from:start,to: start, distance: 0}
        agenda.queue(selfEdge)
        visited[start] = true
        gScore[start] = 0
        const startVertex = {vertexID: start, squareID: this.vertexMap[start]}
        const goalVertex = {vertexID: goal, squareID: this.vertexMap[goal]}
        fScore[start] = heur(startVertex,goalVertex)

        while(agenda.length > 0) {
            const curr = agenda.dequeue()
            this.notifyUI(
                curr.from,
                CURRENT_COLOR
            )
                
            await delay(DELAY)

            // Goal Found
            if(curr.to === goal) {
                console.log("A* SUCCESS, visited node count: ",this.countVisited(visited))
                console.log(this.printPath(goal,parent))
                return
            }

            const neighbors = this.adjList[curr.to]
            for(let neighbor of neighbors) {
                const {from,to,distance} = neighbor
                const tempG = gScore[from] + distance
                if(tempG < gScore[to]) {
                    parent[to] = from
                    gScore[to] = tempG
                    const fromVertex = {vertexID: from, squareID: this.vertexMap[from]}
                    const toVertex = {vertexID: to, squareID: this.vertexMap[to]}
                    fScore[to] = tempG + heur(fromVertex,toVertex)
                    if(!visited[to]) {
                        agenda.queue(neighbor)
                    }
                }
            }

            this.notifyUI(curr.from,VISITED_COLOR)
        }

        console.log("A* No Solution")
    }
}

File: java_src/ai_prj1/Graph.java
package java_src.ai_prj1;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.PriorityQueue;
import java.util.Stack;

public class Graph {
    public static int COLS = 10;
    public static int ROWS = 10;

    private HashMap<Integer, HashSet<Edge>> adjList = new HashMap<>();
    private HashMap<Integer, Integer> vertexMap = new HashMap<>();

    private double[] getSqurePos(int vertexID, int squareID) {
        double[] result = new double[2];

        result[0] = Math.floor(vertexMap.get(vertexID) / ROWS);
        result[1] = squareID / ROWS;

        return result;
    }

    // Eucledean Distance
    private double heur(int a, int b) {
        double[] aPos = getSqurePos(a, vertexMap.get(a));
        double[] bPos = getSqurePos(b, vertexMap.get(b));

        double side1 = Math.abs(aPos[0] - bPos[0]);
        double side2 = Math.abs(aPos[1] - bPos[1]);

        return Math.sqrt(side1*side1 + side2*side2);
    }

    public void addEdge(int from, int to, int distance) {
        Edge edge1 = new Edge(from, to, distance);
        Edge edge2 = new Edge(to, from, distance);

        if(!adjList.containsKey(from))
            adjList.put(from, new HashSet<>());
        if(!adjList.containsKey(to))
            adjList.put(to, new HashSet<>());

        var s1 = adjList.get(from);
        var s2 = adjList.get(to);
        s1.add(edge1);
        s2.add(edge2);
    }

    public void addEdges(List<String[]> edges) {
        for (var edge : edges)
            addEdge(
                Integer.parseInt(edge[0]),
                Integer.parseInt(edge[1]),
                Integer.parseInt(edge[2])
            );
    }

    public void addVertex(int vertexID, int squareID) {
        vertexMap.put(vertexID, squareID);
    }

    public void addVertices(List<String[]> vertices) {
        for (var v : vertices)
            addVertex(Integer.parseInt(v[0]), Integer.parseInt(v[1]));
    }

    private void printPath(int end, int[] parent) {
        Stack<Integer> stack = new Stack<>();

        while(end != -1) {
            stack.push(end);
            end = parent[end];
        }

        ArrayList<Integer> path = new ArrayList<>();
        for(int i = stack.size() - 1; i >= 0; i--)
            path.add(stack.pop());
        
        System.out.println(path);
    }

    public void shortestPath(int start, int goal, String searchType) {
        switch (searchType) {
            case "dijkstra":
                dijkstra(start, goal);
                break;
            case "astar":
                astar(start, goal);
                break;
            default:
                System.out.println("Error, please specify a valid search type");
                break;
        }
    }

    private void dijkstra(int start, int goal) {
        int[] dist = new int[adjList.size()];
        int[] parent = new int[adjList.size()];
        Arrays.fill(dist, Integer.MAX_VALUE);
        Arrays.fill(parent, -1);

        dist[start] = 0;

        PriorityQueue<Edge> agenda = new PriorityQueue<>();
        var selfEdge = new Edge(start, start, 0);
        agenda.add(selfEdge);

        while (!agenda.isEmpty()) {
            Edge currEdge = agenda.poll();
            int currNode = currEdge.getTo();
            if(currEdge.getDistance() > dist[currNode])
                continue;
            
            for(var edge : adjList.get(currNode)) {
                int from = edge.getFrom();
                int to = edge.getTo();
                int distance = edge.getDistance();
                if(dist[currNode] + distance < dist[to]) {
                    dist[to] = dist[currNode] + distance;
                    agenda.add(new Edge(from,to,dist[to]));
                    parent[to] = currNode;
                }
            }
        }

        if(dist[goal] < Integer.MAX_VALUE) {
            System.out.println("Dijkstra Path found with cost: " + dist[goal]);
            printPath(goal, parent);
        }
        else {
            System.out.println("Dijkstra No Solution");
            System.exit(1);
        }
    }

    private void astar(int start, int goal) {
        int len = adjList.size();
        PriorityQueue<Edge> agenda = new PriorityQueue<>();
        boolean[] visited = new boolean[len];
        int[] parent = new int[len];
        double[] gScore = new double[len];
        double[] fScore = new double[len];

        for(int i=0;i<len;i++) {
            visited[i] = false;
            parent[i] = -1;
            gScore[i] = Double.MAX_VALUE;
            fScore[i] = Double.MAX_VALUE;
        }

        var selfEdge = new Edge(start, start, 0);
        agenda.add(selfEdge);
        visited[start] = true;
        gScore[start] = 0;
        fScore[start] = heur(start, goal);

        while(!agenda.isEmpty()) {
            Edge currEdge = agenda.poll();

            // Goal Found
            if(currEdge.getTo() == goal) {
                System.out.println("A* Found Path with cost: "+ gScore[goal]);
                printPath(goal, parent);
                System.exit(1);
            }

            var neighbors = adjList.get(currEdge.getTo());
            for(var neighbor : neighbors) {
                int to = neighbor.getTo();
                int from = neighbor.getFrom();
                int distance = neighbor.getDistance();
                double tempG = gScore[from] + distance;
                if(tempG < gScore[to]) {
                    parent[to] = from;
                    gScore[to] = tempG;
                    fScore[to] = tempG + heur(from, to);
                    if(!visited[to])
                        agenda.add(neighbor);
                }
            }
        }

        // If no solution found fail
        System.out.println("A* didn't find solution");
    }
    
    private class Edge implements Comparable<Edge>{
        private int from;
        private int to;
        private int distance;
    
        public Edge(int from, int to, int distance) {
            this.setFrom(from);
            this.setTo(to);
            this.setDistance(distance);
        }
    
        public int getFrom() {
            return from;
        }
    
        public void setFrom(int from) {
            this.from = from;
        }
    
        public int getDistance() {
            return distance;
        }
    
        public void setDistance(int distance) {
            this.distance = distance;
        }
    
        public int getTo() {
            return to;
        }
    
        public void setTo(int to) {
            this.to = to;
        }

        @Override
        public int compareTo(Edge o) {
            return distance - o.getDistance();
        }
    }
}

